Source/Dafny/Resolver.cs

```diff
@@ -12501,6 +12494,48 @@ private class RBranchExpr : RBranch {
       }
     }
 
+    private ExtendedPattern RemoveDisjunctivePatterns(ExtendedPattern pat, bool allowVariables) {
+      switch (pat) {
+        case LitPattern:
+          return pat;
+        case IdPattern p:
+          if (p.Arguments == null && !p.Id.StartsWith("_") && !allowVariables) {
+            reporter.Error(MessageSource.Resolver, pat.Tok, "Or-patterns may not bind variables");
+            return new IdPattern(p.Tok, FreshTempVarName("_", null), null);
+          }
+          return new IdPattern(p.Tok, p.Id, p.Arguments?.ConvertAll(a => RemoveDisjunctivePatterns(a, allowVariables)));
+        case DisjunctivePattern p:
+          reporter.Error(MessageSource.Resolver, pat.Tok, "Or-patterns are not allowed inside other patterns");
+          return new IdPattern(p.Tok, FreshTempVarName("_", null), null);
+        default:
+          Contract.Assert(false);
+          return null;
+      }
+    }
+
+    private IEnumerable<ExtendedPattern> FlattenDisjunctivePatterns(ExtendedPattern pat) {
+      // TODO: Once we rewrite the pattern-matching compiler, we'll handle disjunctive patterns in it, too.
+      // For now, we handle top-level disjunctive patterns by duplicating the corresponding cases here, and disjunctive
+      // sub-patterns are unsupported.
+      return pat is DisjunctivePattern p
+        ? p.Alternatives.ConvertAll(a => RemoveDisjunctivePatterns(a, allowVariables: false))
+        : Enumerable.Repeat(RemoveDisjunctivePatterns(pat, allowVariables: true), 1);
+    }
+
+    private IEnumerable<NestedMatchCaseExpr> FlattenNestedMatchCaseExpr(NestedMatchCaseExpr c) {
+      var cloner = new Cloner();
+      foreach (var pat in FlattenDisjunctivePatterns(c.Pat)) {
+        yield return new NestedMatchCaseExpr(c.Tok, pat, cloner.CloneExpr(c.Body), cloner.CloneAttributes(c.Attributes));
+      }
+    }
+
+    private IEnumerable<NestedMatchCaseStmt> FlattenNestedMatchCaseStmt(NestedMatchCaseStmt c) {
+      var cloner = new Cloner();
+      foreach (var pat in FlattenDisjunctivePatterns(c.Pat)) {
+        yield return new NestedMatchCaseStmt(c.Tok, pat, c.Body.ConvertAll(cloner.CloneStmt), cloner.CloneAttributes(c.Attributes));
```

Comment:
Why do you need to clone here instead of use the old body and attributes? Will you be using the old cases?

Reply:
Yes, the old body and attributes are used in the linearity checkers.

Reply:
But still, what happens if you don't clone them?

Reply:
Most likely Dafny will crash.  My answer above wasn't complete: notice how I *don't* clone `pat`, because this one isn't being shared.  But I *do* need to clone `c.Body`, because that's the *same* body being copied into each newly created `NestedMatchCaseStmt`.  Basically this code goes from:

```
case A | B | C => body
```

to 

```
case A => Clone(body)
case B => Clone(body)
case C => Clone(body)
```

If we don't clone the body we'll have multiple places in the AST sharing the same reference to the body (and typechecking it multiple times, among other things)

Reply:
Makes sense, thanks!

Reply:
It made sense, but I was wrong :/  The crash I had observed was not due to not cloning, and in fact not cloning is wrong here, because of https://github.com/dafny-lang/dafny/pull/2334 (the key point being that the branches are already partly resolved, and hence cloning them isn't safe).    Thanks for flagging this.